File: app/src/main/java/com/moggot/findmycarlocation/di/component/AppComponent.java
package com.moggot.findmycarlocation.di.component;

import com.moggot.findmycarlocation.App;
import com.moggot.findmycarlocation.di.module.ActivityModule;
import com.moggot.findmycarlocation.di.module.AppModule;
import com.moggot.findmycarlocation.di.module.BillingModule;
import com.moggot.findmycarlocation.di.module.DataModule;
import com.moggot.findmycarlocation.di.module.FragmentModule;
import com.moggot.findmycarlocation.di.module.LocationModule;
import com.moggot.findmycarlocation.di.module.NetworkModule;
import com.moggot.findmycarlocation.di.module.ViewModelModule;

import javax.inject.Singleton;

import dagger.Component;
import dagger.android.AndroidInjector;

@Singleton
@Component(modules = {AppModule.class,
        DataModule.class,
        LocationModule.class,
        NetworkModule.class,
        FragmentModule.class,
        ActivityModule.class,
        ViewModelModule.class,
        BillingModule.class})
interface AppComponent extends AndroidInjector<App> {

    @Component.Builder
    abstract class Builder extends AndroidInjector.Builder<App> {
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/di/component/AppComponent.kt
package com.moggot.findmycarlocation.di.component

import android.content.Context
import com.moggot.findmycarlocation.di.ApplicationProvider
import com.moggot.findmycarlocation.di.module.ContextModule
import com.moggot.findmycarlocation.di.module.DataModule
import com.moggot.findmycarlocation.di.module.LocationModule
import com.moggot.findmycarlocation.di.module.NetworkModule
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(
        modules = [
            ContextModule::class]
)
interface AppComponent : ApplicationProvider {

    class Initializer private constructor() {
        companion object {
            fun init(context: Context): AppComponent {
                return DaggerAppComponent.builder()
                        .contextModule(ContextModule(context))
                        .build()
            }
        }
    }
}
